File: 프로그래머스/Lv.1/160586. 대충 만든 자판/대충 만든 자판.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(vector<string> keymap, vector<string> targets) {
    vector<int> answer;

    vector<int> table(26, 100000);
    for(const auto& v : keymap){
        for(int i = 0 ; i<v.size(); i++){
            table[v[i] - 'A'] = min(table[v[i] - 'A'], i + 1);
        }
    }

    for(const auto& str: targets){
        int total = 0;
        for(const auto c: str){
            if(table[c - 'A'] == 100000){
                total = -1;
                break;
            }else{
                total += table[c-'A'];
            }
        }
        answer.push_back(total);
    }



    return answer;
}

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.*;

class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        Map<Character, Integer> m = new HashMap();
        
        for(int i = 0; i<keymap.length; i++ ){
            String key = keymap[i];
            for(int j = 0; j< key.length(); j++ ){
                char c = key.charAt(j);
                m.put(c, Math.min(m.getOrDefault(c, Integer.MAX_VALUE), j + 1));
            }
        }
        
        int[] ans = new int[targets.length];
        for(int i = 0; i<targets.length; i++){
            int pressCnt = 0;
            boolean isPossible = true;
            for(char c: targets[i].toCharArray()){
                if(!m.containsKey(c)){
                    isPossible = false;
                    break;
                }
                pressCnt+=m.get(c);
            }
            ans[i] = isPossible ? pressCnt : -1;
        }
        
        return ans;

    }
}